diff -iwBEr -U 4 linux-3.10.5/fs/aufs/i_op.c linux-3.10.6/fs/aufs/i_op.c
--- linux-3.10.5/fs/aufs/i_op.c	2013-08-04 19:55:38.000000000 +0000
+++ linux-3.10.6/fs/aufs/i_op.c	2013-08-13 21:56:22.000000000 +0000
@@ -424,9 +424,9 @@
 
 void au_pin_hdir_set_owner(struct au_pin *p, struct task_struct *task)
 {
 #if defined(CONFIG_DEBUG_MUTEXES) || defined(CONFIG_SMP)
-	p->hdir->hi_inode->i_mutex.owner = task;
+	p->task = task;
 #endif
 }
 
 void au_pin_hdir_acquire_nest(struct au_pin *p)
